### PR TITLE
r2r: Use Needleman-Wunsch to diffchar

### PR DESCRIPTION
**Pull request by kazarmy**
**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

The motivations for this pr are twofold:

1. Turn the following char diff and its ilk into something more pleasant to look at:
cmd_pipe-char-diff
1. Drop the dependency on `git diff` for diffcharing.

<ins>Update (diff after)</ins>

after_diff

**Test plan**

Testing has been done on the following testfile:

```
NAME=1
FILE=-
CMDS=<<EOF
wz abcde
ps
EOF
EXPECT=<<EOF
abdde
EOF
RUN

NAME=2
FILE=-
CMDS=<<EOF
wz abcde
ps
EOF
EXPECT=<<EOF
aaaaa
EOF
RUN

NAME=3
FILE=-
CMDS=<<EOF
wz abcde
ps
EOF
EXPECT=<<EOF
eeeee
EOF
RUN

NAME=4
FILE=-
CMDS=<<EOF
wz abcde
ps
EOF
EXPECT=<<EOF
bcdea
EOF
RUN

NAME=5
FILE=-
CMDS=<<EOF
wz abc\ndef\ngh
ps
EOF
EXPECT=<<EOF
abd
gh
def
EOF
RUN

NAME=6
FILE=-
CMDS=<<EOF
wz abc\ndef\ngh
ps
EOF
EXPECT=<<EOF
abdjghidef
EOF
RUN

NAME=7
FILE=-
CMDS=<<EOF
wz abcdefgh
ps
EOF
EXPECT=<<EOF
ijklmnop
EOF
RUN

NAME=8
FILE=-
CMDS=<<EOF
wz abcdefgh
ps
EOF
EXPECT=<<EOF
igklmnop
EOF
RUN

NAME=9
FILE=-
CMDS=<<EOF
wz abcdefgh
# ps
EOF
EXPECT=<<EOF
abcdefgh
EOF
RUN

NAME=10
FILE=-
CMDS=<<EOF
wz abcdefgh
ps
EOF
EXPECT=<<EOF
EOF
RUN

NAME=11
FILE=-
CMDS=<<EOF
wz sub esp, 8\n0x08049029
ps
EOF
EXPECT=<<EOF
sub esp, 9
EOF
RUN

NAME=12
FILE=-
CMDS=<<EOF
wz sub esp, 9\n0x08049029
ps
EOF
EXPECT=<<EOF
sub esp, 9
EOF
RUN

NAME=13
FILE=-
CMDS=<<EOF
wz sub esp, 8\n9999999999
ps
EOF
EXPECT=<<EOF
sub esp, 9
EOF
RUN

NAME=14
# Do not print alignment table for this test
FILE=bins/elf/_Exit (42)
CMDS=#!pipe python3 scripts/get-funcs.py
EXPECT=<<EOF

Function names:
entry0
sym.imp._Exit

Disassembly of entry0:
            ;-- section..text:
            ;-- .text:
            ;-- _start():
/ 16: entry0 ();
|           0x08049020      55             push ebp                    ; [10] -r-x section size 16 named .text
|           0x08049021      89e5           mov ebp, esp
|           0x08049023      83ec08         sub esp, 9
|           0x08049029      6a2a           push 0x2a                   ; '*' ; 42
\           0x0804902b      e8e0ffffff     call sym.imp._Exit          ; void _Exit(int status)

EOF
RUN
```

More tests may be necessary.

**Closing issues**

...

### CODE REVIEW

**Comment by thestr4ng3r** [comment]
I think it would make sense to have all this text diffing somewhere common in RUtil, so it can be reused for radiff2 and stuff.

**Comment by kazarmy** [comment]
Yes but currently it has to work first.

**Comment by trufae** [comment]
The textdiffing code in r_util is ripped from mercurial, which is MIT licensed and written in C. But there's no clean interface api to use it, enotime here to improve that in a reasonable time, too many items in my list right now. iirc i filled an issue about this

File: binr/r2r/r2r.c [hunk 1]
```diff
@@ -532,15 +534,231 @@ static RThreadFunctionRet worker_th(RThread *th) {
...
+				// Print diff
+				// TODO: Handle mismatches
+				int mismatch_pass = 0; // 0, 1 or 2
+				size_t start_mismatch = (size_t)-1;
+				size_t restart_match = (size_t)-1;
+				R2RPrintDiffMode cur_mode = R2R_DIFF_MATCH;
+				R2RCharAlignment cur_align;
+				idx_align = start_align;
+				while (idx_align < 2 * len_expected) {
+					ut8 expected_ch = align_expected[idx_align];
```
Comment (trufae) [inline binr/r2r/r2r.c:696]: const

File: binr/r2r/r2r.c [hunk 2]
```diff
@@ -532,15 +534,231 @@ static RThreadFunctionRet worker_th(RThread *th) {
...
+				size_t start_align = idx_align;
+				// TODO Print alignment (Debug)
+				for (; idx_align < 2 * len_expected; idx_align++) {
+					ut8 ch = align_expected[idx_align];
+					if (align_actual[idx_align] == '\n' && ch != '\n') {
+						printf (ch ? " " : "-");
+					}
+					if (ch == 0) {
+						printf ("-");
+					} else if (ch == '\n') {
```
Comment (trufae) [inline binr/r2r/r2r.c:666]: what about \r and other non printable chars? like > 127?

File: binr/r2r/r2r.c [hunk 3]
```diff
@@ -532,15 +534,231 @@ static RThreadFunctionRet worker_th(RThread *th) {
...
+typedef enum {
+	R2R_DIFF_MATCH, R2R_DIFF_DELETE, R2R_DIFF_INSERT
+} R2RPrintDiffMode;
+
 static void print_diff(const char *actual, const char *expected, bool diffchar) {
 	RDiff *d = r_diff_new ();
 #ifdef __WINDOWS__
 	d->diff_cmd = "git diff --no-index";
 #endif
+	size_t len_expected = strlen (expected);
```
Comment (trufae) [inline binr/r2r/r2r.c:550]: const

File: binr/r2r/r2r.c [hunk 4]
```diff
@@ -532,15 +534,231 @@ static RThreadFunctionRet worker_th(RThread *th) {
...
+					}
+					printf ("%4s ", char_str);
+				}
+				printf ("\n");
+				for (row = 0; row < dim; row++) {
+					if (row && actual[row - 1] == '\n') {
+						char_str[0] = '\\';
+						char_str[1] = 'n';
+					} else {
+						char_str[0] = !row ? ' ' : actual[row - 1];
```
Comment (trufae) [inline binr/r2r/r2r.c:610]: use row? instead of !row?

File: binr/r2r/r2r.c [hunk 5]
```diff
@@ -532,15 +534,231 @@ static RThreadFunctionRet worker_th(RThread *th) {
...
+					}
+					printf ("\n");
+				}
+				// Do alignment
+				size_t idx_expected = len_expected - 1;
+				size_t idx_actual = len_actual - 1;
+				size_t idx_align = 2 * len_expected - 1;
+				size_t pos_row = dim - 1;
+				size_t pos_col = dim - 1;
+				while (pos_row != 0 || pos_col != 0) {
```
Comment (trufae) [inline binr/r2r/r2r.c:625]: !=0 is implicit

File: binr/r2r/r2r.c [hunk 6]
```diff
@@ -532,15 +534,231 @@ static RThreadFunctionRet worker_th(RThread *th) {
...
+					} else {
+						printf ("%c", ch);
+					}
+				}
+				printf ("\n");
+				// Print diff
+				// TODO: Handle mismatches
+				int mismatch_pass = 0; // 0, 1 or 2
+				size_t start_mismatch = (size_t)-1;
+				size_t restart_match = (size_t)-1;
```
Comment (trufae) [inline binr/r2r/r2r.c:691]: both unused?

Comment (kazarmy) [reply binr/r2r/r2r.c]: I originally wanted to use those variables to handle mismatches in the diff-printing state machine (adding a ton of code to it due to combinatorial explosion), but ca85eeb5267b451623d27a928222cfaafb830377 appears to be a far simpler solution that is also correct. So those variables have been removed in ca85eeb5267b451623d27a928222cfaafb830377.

Comment (kazarmy) [reply binr/r2r/r2r.c]: They were there because I didn't want to forget about them.

File: binr/r2r/r2r.c (hunk 2 above)
Comment (kazarmy) [reply binr/r2r/r2r.c:668]: I plan to `#if 0` the code marked `(Debug)`, but to answer the questions:

> what about \r and other non printable chars?

Not sure yet, will probably see how `git diff` handles them.

> like > 127?

Multibyte characters need to be handled in a special manner. Haven't figured out the details.

Right now this pr has been going on for too long and I'm only targeting Minimum Viable Product here, which means ascii-only printable diffs which I guess should suffice for 90-95% of r2r use cases.

Comment (kazarmy) [reply binr/r2r/r2r.c:668]: "Multibyte characters" -> "Multibyte UTF8 characters"

**Comment by trufae** [comment]
Please rebase to make it green

**Comment by trufae** [comment]
Also, i think it's no longer a draft, isnt?

File: binr/r2r/r2r.c (hunk 5 above)
Comment (kazarmy) [reply binr/r2r/r2r.c]: ok d4a002a7e921b5c58617b67e3709c71c82604751

File: binr/r2r/r2r.c (hunk 4 above)
Comment (kazarmy) [reply binr/r2r/r2r.c]: ok c54e8585d44587c436bb72bb299c0989dda8e835

File: binr/r2r/r2r.c (hunk 3 above)
Comment (kazarmy) [reply binr/r2r/r2r.c]: ok ca308b009765b73e07ce203023c5a123ee1e2364

File: binr/r2r/r2r.c (hunk 1 above)
Comment (kazarmy) [reply binr/r2r/r2r.c]: ok ca308b009765b73e07ce203023c5a123ee1e2364

**Comment by kazarmy** [comment]
> Also, i think it's no longer a draft, isnt?

Yes now it isn't a draft. Testfile is in https://github.com/radareorg/radare2/pull/17248#issue-446852851.

File: binr/r2r/r2r.c (hunk 4 above)
Comment (kazarmy) [reply binr/r2r/r2r.c]: also f44311f4835f253063288955ca5a1c69eb40702d

**Comment by kazarmy** [comment]
Updated the description

**Comment by kazarmy** [comment]
This particular score:

```c
const st16 match_nl = 2;
```

can be adjusted later depending on how strong the newline match needs to be, but I think the remaining TODOs need to be resolved first (in other PRs) before it can be set to high values.

**Pull request merged**